design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,706 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:ns-edit:base
+  labels:
+    rbac.crossplane.io/aggregate-to-namespace-edit: "true"
+    rbac.crossplane.io/base-of-namespace-edit: "true"
+rules:
+- apiGroups: [""]
+  resources: [events]
+  verbs: [get, list, watch]
+- apiGroups: [""]
+  resources: [secrets]
+  verbs: ["*"]
+- apiGroups: [apiextensions.crossplane.io]
+  resources: ["*"]
+  verbs: [get, list, watch]
+---
+# This composite resource role aggregates up to any namespace aligned cluster
+# role that declares compatibility by including the XRD label. The same role
+# also aggregates to the crossplane-edit role.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:composite:examplecomposites.xr.example.org:edit
+  labels:
+    rbac.crossplane.io/aggregate-to-crossplane-edit: "true"
+    rbac.crossplane.io/aggregate-to-namespace-edit: "true"
+    rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+rules:
+- apiGroups: [xr.example.org]
+  resources:
+  - exampleclaims
+  - exampleclaims/status
+  # Composite resources are cluster scoped. Binding this cluster role at the
+  # namespace scope will not actually grant access to these composite resources.
+  # Including them in this cluster role allows one crossplane:composite role to
+  # aggregate to both the crossplane-ns-example-edit and crossplane-edit roles.
+ of this role will not grant
+  - examplecomposites
+  - examplecomposites/status
+  verbs: ["*"]
+```
+
+The RBAC manager creates a namespace aligned cluster roles for any namespace
+with one or more annotations named `rbac.crossplane.io/composites.example.org`
+(where `composites.example.org` is the name of an XRD) and with the value
+`xrd-enabled`.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: example
+  annotations:
+    rbac.crossplane.io/composites.example.org: xrd-enabled
+    rbac.crossplane.io/examplecomposites.xr.example.org: xrd-enabled
+```
+
+The above namespace, for example, will result in the creation of the following
+cluster role. Note that the latter two role selectors are derived from the
+annotations of the namespace.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-ns-example-edit
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-namespace-edit: "true"
+      rbac.crossplane.io/base-of-namespace-edit: "true"
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-namespace-edit: "true"
+      rbac.crossplane.io/xrd: composites.example.org
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-namespace-edit: "true"
+      rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+```
+
+### Provider RBAC Mechanics
+
+To create and bind a cluster role for a `Provider` (package):
+
+1. Find its active `ProviderRevision`.
+1. List all CRDs, filter down to those controlled by the active revision.
```

Comment:
Just adding a note from out meeting that I would like for this to be discoverable on the revision itself, but it works either way